File: ch26p758activityView/ActivityViewTest/MyCoolActivity.m


#import "MyCoolActivity.h"

@interface MyCoolActivity()
@property (nonatomic, strong) NSArray* items;
@property (nonatomic, strong) UIImage* image;
@end

@implementation MyCoolActivity

-(id)init {
    self = [super init];
    if (self) {
        UIImage* im = [UIImage imageNamed:@"sunglasses.png"];
        CGFloat largerSize = fmaxf(im.size.height, im.size.width);
        CGFloat scale = 43.0/largerSize;
        CGSize sz = CGSizeMake(im.size.width*scale, im.size.height*scale);
        UIGraphicsBeginImageContextWithOptions(sz, NO, 0);
        [im drawInRect: (CGRect) {CGPointZero, sz}];
        im = UIGraphicsGetImageFromCurrentImageContext();
        UIGraphicsEndImageContext();
        self->_image = im;
    }
    return self;
}

-(NSString *)activityType {
    return @"com.neuburg.matt.coolActivity";
}

-(NSString *)activityTitle {
    return @"Be Cool";
}

-(UIImage *)activityImage {
    return self.image;
}

-(BOOL)canPerformWithActivityItems:(NSArray *)activityItems {
    NSLog(@"can perform %@", activityItems);
    for (id obj in activityItems) {
        if ([obj isKindOfClass: [NSString class]])
            return YES;
    }
    return NO;
}

-(void)prepareWithActivityItems:(NSArray *)activityItems {
    NSLog(@"prepare %@", activityItems);
    self.items = activityItems;
}

-(void)performActivity {
    NSLog(@"performing %@", self.items);
    [self activityDidFinish:YES];
}


@end

File: iOS9bookExamples/bk2ch13p635activityView/ch26p901activityView/MyCoolActivity.swift

import UIKit
func imageOfSize(size:CGSize, _ opaque:Bool = false, _ closure:() -> ()) -> UIImage {
    UIGraphicsBeginImageContextWithOptions(size, opaque, 0)
    closure()
    let result = UIGraphicsGetImageFromCurrentImageContext()
    UIGraphicsEndImageContext()
    return result
}

class MyCoolActivity : UIActivity {
    var items : [AnyObject]?
    var image : UIImage
    
    override init() {
        let idiom = UIScreen.mainScreen().traitCollection.userInterfaceIdiom
        var scale : CGFloat = (idiom == .Pad ? 76 : 60) - 10
        let im = UIImage(named:"sunglasses.png")!
        let largerSize = fmax(im.size.height, im.size.width)
        scale /= largerSize
        let sz = CGSizeMake(im.size.width*scale, im.size.height*scale)
        self.image = imageOfSize(sz) {
            im.drawInRect(CGRect(origin: CGPoint(), size: sz))
        }
        super.init()
    }
    
    override class func activityCategory() -> UIActivityCategory {
        return .Action // the default
    }
    
    override func activityType() -> String? {
        return "com.neuburg.matt.coolActivity"
    }
    
    override func activityTitle() -> String? {
        return "Be Cool"
    }
    
    override func activityImage() -> UIImage? {
        return self.image
    }
    
    override func canPerformWithActivityItems(activityItems: [AnyObject]) -> Bool {
        print("cool can perform \(activityItems)")
        for obj in activityItems {
            if obj is String {
                print("returning true")
                return true
            }
        }
        print("returning false")
        return false
    }
    
    override func prepareWithActivityItems(activityItems: [AnyObject]) {
        print("cool prepare \(activityItems)")
        self.items = activityItems
    }
    
    override func performActivity() {
        print("cool performing \(self.items)")
        self.activityDidFinish(true)
    }
    
    deinit {
        print("cool activity dealloc")
    }

}